usb/device/targets/TARGET_STM/USBPhy_STM32.cpp

```diff
@@ -0,0 +1,522 @@
+            {0                     | USB_EP_ATTR_DIR_IN_AND_OUT,  1, 4},
+            {0                     | USB_EP_ATTR_DIR_IN_AND_OUT,  1, 4}
+        }
+    };
+    return &table;
+}
+
+uint32_t USBPhyHw::ep0_set_max_packet(uint32_t max_packet)
+{
+    // FUTURE - set endpoint 0 size and return this size
+    return MAX_PACKET_SIZE_EP0;
+}
+
+// read setup packet
+void USBPhyHw::ep0_setup_read_result(uint8_t *buffer, uint32_t size)
+{
+    if (size > MAX_PACKET_SIZE_SETUP) {
+        size = MAX_PACKET_SIZE_SETUP;
+    }
+    memcpy(buffer,  hpcd.Setup, size);
+    memset(hpcd.Setup,0,MAX_PACKET_SIZE_SETUP);
+}
+
+void USBPhyHw::ep0_read(uint8_t *data, uint32_t size)
+{
+    _read_buf[0] = data;
+    _read_size[0] = size;
+
+    uint8_t *pBuf = (uint8_t *)pBufRx0;
+    HAL_StatusTypeDef ret;
+    epComplete[EP_TO_IDX(EP0OUT)] = 2;
+    ret = HAL_PCD_EP_Receive(&hpcd, EP0OUT, pBuf, MAX_PACKET_SIZE_EP0 );
+    MBED_ASSERT(ret!=HAL_BUSY);
+}
+
+uint32_t USBPhyHw::ep0_read_result()
+{
+    uint32_t length = (uint32_t) HAL_PCD_EP_GetRxCount(&hpcd, 0);
+    epComplete[EP_TO_IDX(EP0OUT)] = 0;
+    if (length) {
+        uint8_t  *buff = (uint8_t *)pBufRx0;
+        memcpy(_read_buf[0], buff, _read_size[0] > length ? length : _read_size[0]);
+    }
+    _read_buf[0] = 0;
+    _read_size[0] = 0;
+    return length;
+}
+
+void USBPhyHw::ep0_write(uint8_t *buffer, uint32_t size)
+{
+    /*  check that endpoint maximum size is not exceeding TX fifo */
+    MBED_ASSERT(hpcd.IN_ep[0].maxpacket >= size);
+    endpoint_write(EP0IN, buffer, size);
+}
+
+void USBPhyHw::ep0_stall()
+{
+    endpoint_stall(EP0IN);
+    endpoint_stall(EP0OUT);
+}
+
+bool USBPhyHw::endpoint_add(usb_ep_t endpoint, uint32_t max_packet, usb_ep_type_t type)
+{
+    uint32_t len;
+    if (max_packet > MAXTRANSFER_SIZE) return false;
+    if (endpoint & 0x80) {
+        HAL_PCDEx_SetTxFiFo(&hpcd, endpoint & 0x7f, (max_packet / 4) + 1);
+        len = HAL_PCDEx_GetTxFiFo(&hpcd,endpoint & 0x7f);
+        MBED_ASSERT(len >= max_packet);
+    }
+    HAL_StatusTypeDef ret = HAL_PCD_EP_Open(&hpcd, endpoint, max_packet, type);
+    MBED_ASSERT(ret!=HAL_BUSY);
+    return (ret == HAL_OK) ? true:false;
+}
+
+void USBPhyHw::endpoint_remove(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret = HAL_PCD_EP_Close(&hpcd, endpoint);
+    MBED_ASSERT(ret == HAL_OK);
+}
+
+void USBPhyHw::endpoint_stall(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret;
+    ret = HAL_PCD_EP_SetStall(&hpcd, endpoint);
+    MBED_ASSERT(ret!=HAL_BUSY);
+}
+
+void USBPhyHw::endpoint_unstall(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret;
+    ret = HAL_PCD_EP_ClrStall(&hpcd, endpoint);
+    MBED_ASSERT(ret!=HAL_BUSY);
+}
+
+bool USBPhyHw::endpoint_read(usb_ep_t endpoint, uint8_t *data, uint32_t size)
+{
+    _read_buf[EP_TO_LOG(endpoint)] = data;
+    _read_size[EP_TO_LOG(endpoint)] = size;
+
+    uint8_t* pBuf = (uint8_t *)pBufRx; //TODO - this buffer shouldn't be shared for multiple endpoints
```

Comment:
The same buffer is used for all endpoints which will corrupt data. Any way this can be done with the buffer passed in rather than a dedicated RX buffer?

Reply:
The dummer buffer pBuffRx is there mainly because, in previous API ,the read function does not have buffer in parameter. Only the read result function was having a buffer. Now with the new API , it is the opposite. So the buffer (data) can be used instead of the dummy buffer pBufRx  that was used for any request.

Reply:
@jamike I have removed these buffers and am now passing data directly into the ST hal directly. There aren't any alignment requirements on the data is there?